docs/docs/settings.md
# Settings

The NSClient++ settings store is a hierarchical tree structure with key value pairs.
Traditionally this is stored in a flat ini-style file where the "paths" are defined as sections.
But there are other options as well and for instance the registry is another common place to store settings.
With the new configuration UI in 0.4.3 it is simpler to manipulate the settings if they are store in the registry and thus probably a better place to keep them.

Which keys are available and what they mean are describe by the modules which use the various keys.
Thus it the place to find the documentation for the various configuration options are in the reference section of the documentation.
Where it is split by module.

## Include

Another really useful feature of the settings in NSClient++ is the ability to include various other settings.
This is very flexible and you can include ini file from the registry and vice versa.

The way to include a file (if you are using ini-files) is to add a key under the /includes section.

including file:

```
[/includes]
foo=foo.ini
```

You can include any number of files registry or other stores. and they will be instantiated in a tree structure with a parent child relationship.
Important to note here is that the first found key will be used. So parents will override children.

And example of this:

*   nsclient.ini:
```
[/includes]
client=client.ini

[/test]
key1=This values comes from nsclient.ini
```
*   client.ini:
```
[/includes]
baseline=baseline.ini

[/test]
key1=This values comes from client.ini
key2=This values comes from client.ini
```
*   baseline.ini:
```
[/test]
key1=This values comes from baseline.ini
key2=This values comes from baseline.ini
key3=This values comes from baseline.ini
```

in the above example the values of /test keyx will be:

-   key1=This values comes from nsclient.ini
-   key2=This values comes from client.ini
-   key3=This values comes from baseline.ini

This can be very useful to distribute a baseline configuration for a company or monitoring product.
Then all "machine specific customization" would go into the nsclient.ini config where as client.ini would be reserved for the clients global config. And finally baseline.ini would be monitoring tool specific configuration.


## Settings stores

Settings are historically stored in an ini file but you can store settings in many other locations as well.
While the most obvious one to use is the registry there are other options as well.

-   ini files
-   old ini files
-   registry
-   dummy
-   http (ini files over http)


### Ini settings

Ini file are the simplest form of configuration and also the default though on windows registry is probably a better option.
The files are text-files following the ini file format where you have sections in brackets [] and key key = values.

sample.ini:

```
[/section/child/section]
key=value
```

To use an ini file you prefix the settings url with ini:// then you can use various folder strings or specify a relative or absolute path to the file.

examples:

*   Use the relative file foo.ini `nscp settings --migrate-to ini://foo.ini`
*   Use the relative file foo.ini in a subfolder called conf `nscp settings --migrate-to ini://conf/foo.ini`
*   Use a file stored in the profile folder (on windows): C:\Documents and Settings\All Users\Application Data\NSClient++ `nscp settings --migrate-to ini://%(common-appdata)/NSClient++/nsclient.ini`

### registry settings

Registry is only available on windows and using them on windows is recommended as it integrates better with windows and windows management tools where you can push configuration changes and similar things.
As the registry is naturally a tree structure we use folders as section and keys and values for keys and values.

To use an ini file you prefix the settings url with ini:// then you can use various folder strings or specify a relative or absolute path to the file.

examples:

```
# Use the default registry location
nscp settings --migrate-to registry
# Use HKEY_LOCAL_MACHINE/software/NSClient++ to store configuration
nscp settings --migrate-to registry://HKEY_LOCAL_MACHINE/software/NSClient++
```

### http settings

Http/Https is a regular ini file (see above) except that it is loaded remotely and refreshed periodically.
The refresh period is configurable and if the file has changed NSClient++ will reload and re-read the new configuration.
If it is not possible to fetch the latest configuration NSClient++ will keep using the last one it received.

examples:
```
# Use remote configuration.
nscp settings --switch http://www.myserver.com/nsclient.ini
```

In the nsclient.ini file you can specify a series of attachments which will be downloaded (for instance scripts).

Adding a script:

```
[/attachments]
scripts/myscript.bat = http://www.myserver.com/myscript.bat
```

#### Using TLS

You likely want to use TLS when using http settings.
To use TLS you need to configure either If you are using a custom CA or want to use self signed certificates that has to be configured in the `boot.ini` file.
This is a special file which is used to load configuration before the configuration is loaded.
For instance, it defines which settings store to use but you can also configure TLS options.

> It is advised to set `verify mode` to `peer` and use a CA certificate to verify the server certificate.
> This will become the default in the future.

```ini
[tls]
version=1.3
verify mode=peer
ca=c:\program files\NSClient++\security\ca.pem
```

| Key         | Default Value | Values          | Description                                                                |
|-------------|---------------|-----------------|----------------------------------------------------------------------------|
| version     | 1.3           | 1.0, 1.1, 1.3   | The TLS version to use.                                                    |
| verify mode | none          | none, peer      | The verify mode to use (Set this to none to use self signed certificates). |
| ca          |               | Path to CA file | The path to the CA certificate to use.                                     |

## Using settings stores

NSClient++ has some feature to help work with settings stores.

All of this is performed via the settings sub command: `nscp settings --help`

Migrating settings from a ini file to the registry can be done with the migrate-to command: `nscp settings --migrate-to registry`

This will move all data from the settings file and store it in in the registry and then setup NSClient++ to use the registry instead of the ini file.

You can also switch settings store (without migrating data): `nscp settings --switch registry`

The effect is similar in that NSClient++ will start using the registry but you have to add the keys to the registry manually.

To show the current settings store view can run:
```
$ nscp settings --show
INI settings: (ini://${shared-path}/nsclient.ini, C:\source\build\x64\dev/nsclient.ini)
```

## Overriding settings store

The default way for NSClient++ to find your settings files is via the boot.ini file.

If you want to override this, for instance you want to use multiple NSClient++ from the same folder, you can do this via the `--settings` option: `nscp test --settings nsclient2.ini`

You can do this for the service as well by editing the service start command.

## Paths

Paths can be used in various places in the settings store to locate files. To facilitate reusable paths there a number of pat variables which ca be used.

| Key              | Value (Windows)                 | Value (Linux)          | Comment                                                                |
|------------------|---------------------------------|------------------------|------------------------------------------------------------------------|
| certificate-path | ${shared-path}/security         |                        |                                                                        |
| module-path      | ${exe-path}/modules             | ${shared-path}/scripts |                                                                        |
| web-path         | ${shared-path}/web              |                        |                                                                        |
| scripts          | ${exe-path}/scripts             | ${shared-path}/scripts |                                                                        |
| cache-folder     | ${shared-path}/cache            |                        |                                                                        |
| crash-folder     | ${shared-path}/crash-dumps      |                        |                                                                        |
| log-path         | ${shared-path}/log              |                        |                                                                        |
| base-path        | Path of NSClient++ exe file     |                        | This will in the future change to an actual shared path.               |
| temp             | The temporary file path         | /tmp                   |                                                                        |
| shared-path      | Path of NSClient++ exe file     |                        | This will in the future change to an actual shared path.               |
| exe-path         | Path of NSClient++ exe file     |                        |                                                                        |
| common-appdata   | Application data for all users. | N/A                    | The file system directory that contains application data for all users |
| appdata          | The user's profile folder.      | N/A                    |                                                                        |
| etc              | N/A                             | /etc                   | Linux only                                                             |

All paths can also be overridden using the /paths section in the configuration file.
Please note that that is not always possible for instance overriding shared-path and reading settings from shared path will not work.
Paths overrides are only available AFTER the configuration has been read.

Example of overriding a path (web root folder).

```
[/paths]
web-path=/tmp/foo
```

## Security

Storing sensitive information in the settings store is not recommended.
You can solve this in various ways: 
 * The simplest approach is to move the settings file to a location only readable by the user running NSClient++.
 * Use the credential manager to store sensitive information.

To learn more about the credential manager see [Securing NSClient++](securing.md).

docs/manual/nscp_settings.rst
*******************
Using settings mode
*******************

Used for manipulating the settings store in any conceivable way. Everything from switching context to migrating settings to settings keys and generating documentation.

changelog
Hence forth change log can be found on github: https://github.com/mickem/nscp/releases

2023-07-30 Michael Medin
 * Added first version of github action build pipeline

2018-02-03 Michael Medin
 * Updated documentation (#513)
 * eventlog: Fixed issue where real-time eventlog message return duplicate results if same log with different case was added
 * NSCA Client: Fixed #511 nsca status can be set with a number from command line
 * filters: show-all will not change if the list is already present: fixes #528
 * Added dockerfiles for debian #537
 * Some linux build fixes
 * Fixed 401 issue with Op5Client
 * corrected spelling

2018-01-18 Michael Medin
 * Fixed some Op5Client issues

2018-01-17 Philipp
 * Fix syntax error for PowerShell wrapper

 2018-01-15 Michael Medin
 * Added hidden to check_tasksched to allow checking of hidden tasks
 * Added tracing and fixed some issues to op5 client
 * Fixed #525 json spirit should be an optional dependency (though a lot of things break without it)

2018-01-08 Michael Medin
 * Possibly fixed #525 json-spirit should not be a required dependancy
 * Fixed #527 infinite recursion with malformed config file
 * Fixed #518 silent installer not applying cli values
 * Fixed so OP5_SERVER et.al. are applied correctly to the installer
 * Added new OP5Client
 * optimized the markdown a bit
 * started to fix debian builds
 * Added http client for mongoose wrapper
 * moved moongose impl behind the c++ wrappers
 * add centreon plugins to docs: ref #512

2017-11-19 Michael Medin
 * REST: Added crude settings API (needs more work before making official)
 * Fixed #435 added support for overriding http verb with X-HTTP-Method-Override
 * Improved error message for 404:s
 * Fixed some issues with enumerating settings keys
 * Removed query from settings protobuf api as it was never used
 * CORE: Fixed listing unloaded modules by name
 * REST: Added POST of modules to upload them
 * REST: Fixed execute_nagios support for plain text output and fixed return code
 * REST: Fixed segv when parsing privileges
 * WEBServer: Fixed output from nscp web install
 * zip-modules: fixed loading zip-modules without extensions
 * REST: GET ...queries/ actually lists scripts not queries
 * REST: Added putting log messages via REST API.
 * Spellchecked some docs a bit

2017-11-16 Michael Medin
 * Added more python API documentation
 * Fixed #499 Added details about adding performance data to documentation about checks
 * Fixed #502 removed pointless log message about logger
 * REST: Fixed #485 check_nscp_web is now supported again, sorry for the delay....

2017-11-15 Michael Medin
 * Added command line commands to nscp web for adding users and roles: nscp web <add-user|add-role> --help
 * REST: Added support for python runtime

2017-11-14 Michael Medin
 * REST: Added support for loading zip modules
 * REST: Fixed fetching single plugin and plugin load status updates
 * REST: fixed segv when fetching log records
 * REST: Added links to modules controller
 * Fixed python script CLI for loading scripts
 * Added some more python API docs
 * Removed erroneous error message bout web password
 * Added compatibility matrices to documentation (Fixing #482 )

2017-11-12 Michael Medin
 * check_process: Fixed #494 empty status message
 * Events: Added proper event names
 * CheckEventLog added support for fetching xml
 * REST: added privileges to query controller
 * Fixed rendering issue in api docs

2017-11-11 Michael Medin
 * Added skeleton docs for the Python API
 * Improved the markdown  docs for the plugin API
 * Various minor rest api fixes (pagination, and links)
 * Fixed python script issues

2017-11-09 Michael Medin
 * Fixed accidental renaming of module in ext-scr install command
 * Fixed so clients (test/web) can show multiple lines in output

2017-11-08 Michael Medin
 * Added info API to get version

2017-11-06 Michael Medin
 * Regression fixes for external scripts
 * Fixed some python script refactoring regression issues

2017-11-05 Michael Medin
 * check_eventlog: Added support for audit success/failure keyword filtering Fixed #488
 * zip modules: added support for python script one-liners
 * PythonScripts: Made python scripts compatible with external scripts cli as well as code refactoring
 * check_pdh: Added ignore-errors to make counters return 0 instead of errors: #454
 * Split external script up into chunks

2017-10-31 Michael Medin
 * ZIP modules: Added modules and on_start to zip modules to load modules and run commands on start

2017-10-28 Michael Medin
 * Implemented script support for zip plugins

2017-10-28 Michael Medin
 * CheckExternalScripts added support for not writing the configuration when adding script (--no-config)
 * Added ZIP plugins

2017-10-18 Thomas Rohlajz
 * Remove duplicity in tutorial documentation

2017-10-25 Michael Medin
 * REST: added log api
 * core: Refactoring

2017-10-17 Michael Medin
 * Fixed #472 Added sample exclude versus filter
 * Fixed #471 added sample command with perf-config to check_memory

2017-10-16 Michael Medin
 * Fixed #413 check_network returning the wrong values
 * Added example real-time eventlog command to docs

2017-10-15 Michael Medin
 * Major overhaul of the documentation making the generated bits easier to read...

2017-10-11 Michael Kraus
 * Add link and information about check_nsc_web

2017-10-11 Michael Medin
 * Added documentation for REST API and added privs and fixed some issues

2017-10-10 Michael Medin
 * Updated the API documentation
 * Added /api and /api/v1 endpoints to provide information about API
 * Added API documentation for scripts
 * Added configuration for grants and roles #448 #449
 * Added privilege system for API #448 and #449
 * Added support for basic auth (#448)

2017-10-05 Michael Medin
 * Fixed installer issue
 * Renamed the moongose lib so it wont clash with other thing on linux

2017-10-01 Michael Medin
 * implemented DELETE for /api/v1/scripts/ext/scripts to delete commands and scripts.

include/config.h
/**************************************************************************
*   Copyright (C) 2004-2007 by Michael Medin <michael@medin.name>         *
*                                                                         *
*   This code is part of NSClient++ - http://trac.nakednuns.org/nscp      *
*                                                                         *
*   This program is free software; you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation; either version 2 of the License, or     *
*   (at your option) any later version.                                   *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program; if not, write to the                         *
*   Free Software Foundation, Inc.,                                       *
*   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
***************************************************************************/
#pragma once

#include "../AutoBuild.h"
// Application Name
#define SZAPPNAME _T("NSClient++")

// Version
#define SZBETATAG _T(" ")
//#define SZBETATAG _T(" RC ")  
//#define SZBETATAG _T(" BETA ") 
#define SZVERSION STRPRODUCTVER SZBETATAG STRPRODUCTDATE
//FILEVER[0]

#if defined(_M_IX86)
#define SZARCH _T("w32")
#elif defined(_M_X64)
#define SZARCH _T("x64")
#elif defined(_M_IA64)
#define SZARCH _T("ia64")
#else
#define SZARCH _T("unknown")
#endif

// internal name of the service
#define SZSERVICENAME        _T("NSClientpp")

// Description of service
#define SZSERVICEDESCRIPTION _T("Nagios Windows Agent (Provides performance data for Nagios server)")

// displayed name of the service
#define SZSERVICEDISPLAYNAME SZSERVICENAME _T(" (Nagios) ") SZVERSION _T(" ") SZARCH

// list of service dependencies - "dep1\0dep2\0\0"
#define SZDEPENDENCIES       _T("")

// Buffer size of incoming data (notice this is the maximum request length!)
#define RECV_BUFFER_LEN		1024

#define NASTY_METACHARS         _T("|`&><'\"\\[]{}")        /* This may need to be modified for windows directory seperator */

#define DATE_FORMAT _T("%#c")


// Default Argument string (for consistency)
#define SHOW_ALL _T("ShowAll")
#define SHOW_FAIL _T("ShowFail")
#define NSCLIENT _T("nsclient")
#define IGNORE_PERFDATA _T("ignore-perf-data")
#define CHECK_ALL _T("CheckAll")
#define CHECK_ALL_OTHERS _T("CheckAllOthers")

// NSClient Setting headlines
#define NSCLIENT_SECTION_TITLE _T("NSClient")
#define NSCLIENT_SETTINGS_PORT _T("port")
#define NSCLIENT_SETTINGS_PORT_DEFAULT 12489
#define NSCLIENT_SETTINGS_VERSION _T("version")
#define NSCLIENT_SETTINGS_VERSION_DEFAULT _T("auto")
#define NSCLIENT_SETTINGS_BINDADDR _T("bind_to_address")
#define NSCLIENT_SETTINGS_BINDADDR_DEFAULT _T("")
#define NSCLIENT_SETTINGS_LISTENQUE _T("socket_back_log")
#define NSCLIENT_SETTINGS_LISTENQUE_DEFAULT 0
#define NSCLIENT_SETTINGS_READ_TIMEOUT _T("socket_timeout")
#define NSCLIENT_SETTINGS_READ_TIMEOUT_DEFAULT 30
#define NSCLIENT_SETTINGS_SYSTRAY_EXE _T("systray_exe")
#define NSCLIENT_SETTINGS_SYSTRAY_EXE_DEFAULT _T("nstray.exe")

// NRPE Settings headlines
#define NRPE_SECTION_TITLE _T("NRPE")
#define NRPE_CLIENT_HANDLER_SECTION_TITLE _T("NRPE Client Handlers")
#define NRPE_HANDLER_SECTION_TITLE _T("NRPE Handlers")
#define NRPE_SETTINGS_TIMEOUT _T("command_timeout")
#define NRPE_SETTINGS_TIMEOUT_DEFAULT 60
#define NRPE_SETTINGS_READ_TIMEOUT _T("socket_timeout")
#define NRPE_SETTINGS_READ_TIMEOUT_DEFAULT 30
#define NRPE_SETTINGS_PORT _T("port")
#define NRPE_SETTINGS_PORT_DEFAULT 5666
#define NRPE_SETTINGS_BINDADDR _T("bind_to_address")
#define NRPE_SETTINGS_BINDADDR_DEFAULT _T("")
#define NRPE_SETTINGS_ALLOW_ARGUMENTS _T("allow_arguments")
#define NRPE_SETTINGS_ALLOW_ARGUMENTS_DEFAULT 0
#define NRPE_SETTINGS_ALLOW_NASTY_META _T("allow_nasty_meta_chars")
#define NRPE_SETTINGS_ALLOW_NASTY_META_DEFAULT 0
#define NRPE_SETTINGS_USE_SSL _T("use_ssl")
#define NRPE_SETTINGS_USE_SSL_DEFAULT 1
#define NRPE_SETTINGS_LISTENQUE _T("socket_back_log")
#define NRPE_SETTINGS_LISTENQUE_DEFAULT 0
#define NRPE_SETTINGS_PERFDATA _T("performance_data")
#define NRPE_SETTINGS_PERFDATA_DEFAULT 1
#define NRPE_SETTINGS_SCRIPTDIR _T("script_dir")
#define NRPE_SETTINGS_SCRIPTDIR_DEFAULT _T("")
#define NRPE_SETTINGS_STRLEN _T("string_length")
#define NRPE_SETTINGS_STRLEN_DEFAULT 1024
#define NRPE_SETTINGS_COUNT _T("packet_count")
#define NRPE_SETTINGS_COUNT_DEFAULT 1

// External Script Settings headlines
#define EXTSCRIPT_SECTION_TITLE _T("External Script")
#define EXTSCRIPT_SCRIPT_SECTION_TITLE _T("External Scripts")
#define EXTSCRIPT_ALIAS_SECTION_TITLE _T("External Alias")
#define EXTSCRIPT_SETTINGS_TIMEOUT _T("command_timeout")
#define EXTSCRIPT_WRAPPINGS_SECTION_TITLE _T("Script Wrappings")
#define EXTSCRIPT_WRAPPED_SCRIPT_SECTION_TITLE _T("Wrapped Scripts")
#define EXTSCRIPT_SETTINGS_TIMEOUT_DEFAULT 60
#define EXTSCRIPT_SETTINGS_ALLOW_ARGUMENTS _T("allow_arguments")
#define EXTSCRIPT_SETTINGS_ALLOW_ARGUMENTS_DEFAULT 0
#define EXTSCRIPT_SETTINGS_ALLOW_NASTY_META _T("allow_nasty_meta_chars")
#define EXTSCRIPT_SETTINGS_ALLOW_NASTY_META_DEFAULT 0
#define EXTSCRIPT_SETTINGS_SCRIPTDIR _T("script_dir")
#define EXTSCRIPT_SETTINGS_SCRIPTDIR_DEFAULT _T("")

// Check System Settings
#define C_SYSTEM_SECTION_TITLE _T("Check System")
#define C_SYSTEM_CPU_BUFFER_TIME _T("CPUBufferSize") 
#define C_SYSTEM_CPU_BUFFER_TIME_DEFAULT _T("1h")
#define C_SYSTEM_CHECK_RESOLUTION _T("CheckResolution")
#define C_SYSTEM_CHECK_RESOLUTION_DEFAULT 10 /* unit: second/10 */

#define C_SYSTEM_AUTODETECT_PDH _T("auto_detect_pdh")
#define C_SYSTEM_AUTODETECT_PDH_DEFAULT 1
#define C_SYSTEM_FORCE_LANGUAGE _T("force_language")
#define C_SYSTEM_FORCE_LANGUAGE_DEFAULT _T("")
#define C_SYSTEM_NO_INDEX _T("dont_use_pdh_index")
#define C_SYSTEM_NO_INDEX_DEFAULT 0
#define C_SYSTEM_IGNORE_COLLECTION _T("debug_skip_data_collection")
#define C_SYSTEM_IGNORE_COLLECTION_DEFAULT 0
#define C_SYSTEM_PDH_SUBSYSTEM _T("pdh_subsystem")
#define C_SYSTEM_PDH_SUBSYSTEM_DEFAULT _T("fast")

#define C_SYSTEM_MEM_PAGE_LIMIT _T("MemoryCommitLimit")
#define C_SYSTEM_MEM_PAGE_LIMIT_DEFAULT _T("\\Memory\\Commit Limit")
#define C_SYSTEM_MEM_PAGE _T("MemoryCommitByte")
#define C_SYSTEM_MEM_PAGE_DEFAULT _T("\\Memory\\Committed Bytes")
#define C_SYSTEM_UPTIME _T("SystemSystemUpTime")
#define C_SYSTEM_UPTIME_DEFAULT _T("\\System\\System Up Time")
#define C_SYSTEM_CPU _T("SystemTotalProcessorTime")
#define C_SYSTEM_MEM_CPU_DEFAULT _T("\\Processor(_total)\\% Processor Time")
#define C_SYSTEM_ENUMPROC_METHOD_PSAPI _T("PSAPI")
#define C_SYSTEM_ENUMPROC_METHOD_THELP _T("TOOLHELP")
#define C_SYSTEM_ENUMPROC_METHOD_AUTO _T("auto")
#define C_SYSTEM_ENUMPROC_METHOD _T("ProcessEnumerationMethod")
#define C_SYSTEM_ENUMPROC_METHOD_DEFAULT C_SYSTEM_ENUMPROC_METHOD_AUTO

#define CHECK_DISK_SECTION_TITLE _T("Check Disk")
#define CHECK_DISK_SHOW_ERRORS _T("show_errors")
#define CHECK_DISK_SHOW_ERRORS_DEFAULT 0

#define EVENTLOG_SECTION_TITLE _T("Eventlog")
#define EVENTLOG_DEBUG _T("debug")
#define EVENTLOG_DEBUG_DEFAULT 0
#define EVENTLOG_LOOKUP_NAMES _T("lookup_names")
#define EVENTLOG_LOOKUP_NAMES_DEFAULT 1
#define EVENTLOG_SYNTAX _T("syntax")
#define EVENTLOG_SYNTAX_DEFAULT _T("%severity%: %source%: %message% (%count%)")
#define EVENTLOG_BUFFER _T("buffer_size")
#define EVENTLOG_BUFFER_DEFAULT 1024*64

#define NSCA_AGENT_SECTION_TITLE _T("NSCA Agent")
#define NSCA_CMD_SECTION_TITLE _T("NSCA Commands")
#define NSCA_INTERVAL _T("interval")
#define NSCA_INTERVAL_DEFAULT 60
#define NSCA_HOSTNAME _T("hostname")
#define NSCA_HOSTNAME_DEFAULT _T("")
#define NSCA_SERVER _T("nsca_host")
#define NSCA_SERVER_DEFAULT _T("unknown-host")
#define NSCA_PORT _T("nsca_port")
#define NSCA_PORT_DEFAULT 5667
#define NSCA_ENCRYPTION _T("encryption_method")
#define NSCA_ENCRYPTION_DEFAULT 1
#define NSCA_PASSWORD _T("password")
#define NSCA_PASSWORD_DEFAULT _T("")
#define NSCA_DEBUG_THREADS _T("debug_threads")
#define NSCA_DEBUG_THREADS_DEFAULT 1
#define NSCA_CACHE_HOST _T("cache_hostname")
#define NSCA_CACHE_HOST_DEFAULT 0
#define NSCA_TIME_DELTA _T("time_delay")
#define NSCA_TIME_DELTA_DEFAULT _T("0")
#define NSCA_REPORT _T("report")
#define NSCA_REPORT_DEFAULT _T("all")
#define NSCA_READ_TIMEOUT _T("socket_timeout")
#define NSCA_READ_TIMEOUT_DEFAULT 30
#define NSCA_STRLEN _T("string_length")
#define NSCA_STRLEN_DEFAULT 512
#define NSCA_CHUNK _T("chunk")
#define NSCA_CHUNK_DEFAULT 0


#define C_SYSTEM_SVC_ALL_0 _T("check_all_services[SERVICE_BOOT_START]")
#define C_SYSTEM_SVC_ALL_0_DEFAULT _T("ignored")
#define C_SYSTEM_SVC_ALL_1 _T("check_all_services[SERVICE_SYSTEM_START]")
#define C_SYSTEM_SVC_ALL_1_DEFAULT _T("ignored")
#define C_SYSTEM_SVC_ALL_2 _T("check_all_services[SERVICE_AUTO_START]")
#define C_SYSTEM_SVC_ALL_2_DEFAULT _T("started")
#define C_SYSTEM_SVC_ALL_3 _T("check_all_services[SERVICE_DEMAND_START]")
#define C_SYSTEM_SVC_ALL_3_DEFAULT _T("ignored")
#define C_SYSTEM_SVC_ALL_4 _T("check_all_services[SERVICE_DISABLED]")
#define C_SYSTEM_SVC_ALL_4_DEFAULT _T("stopped")
#define C_SYSTEM_SVC_ALL_5 _T("check_all_services[SERVICE_DELAYED]")
#define C_SYSTEM_SVC_ALL_5_DEFAULT _T("ignored")

#define C_TASKSCHED_SECTION _T("Task Scheduler")
#define C_TASKSCHED_SYNTAX _T("syntax")
#define C_TASKSCHED_SYNTAX_DEFAULT _T("%title% last run: %most-recent-run-time% (%exit-code%)")

// Log to File Settings
#define LOG_SECTION_TITLE _T("log")
#define LOG_FILENAME _T("file") 
#define LOG_FILENAME_DEFAULT _T("nsclient.log")
#define LOG_DATEMASK _T("date_mask")
#define LOG_DATEMASK_DEFAULT _T("%Y-%m-%d %H:%M:%S")
#define LOG_ROOT _T("root_folder")
#define LOG_ROOT_DEFAULT _T("exe")


// CRash settings
#define CRASH_SECTION_TITLE _T("crash")
#define CRASH_RESTART _T("restart")
#define CRASH_RESTART_DEFAULT 1
#define CRASH_RESTART_NAME _T("service_name")

#define CRASH_SUBMIT _T("submit")
#define CRASH_SUBMIT_DEFAULT 0
#define CRASH_SUBMIT_URL _T("url")
#define CRASH_SUBMIT_URL_DEFAULT _T("http://crash.nsclient.org/submit")

#define CRASH_ARCHIVE _T("archive")
#define CRASH_ARCHIVE_DEFAULT 1
#define CRASH_ARCHIVE_FOLDER _T("folder")

// Main Settings
#define MAIN_SECTION_TITLE _T("Settings")
#define MAIN_USEFILE _T("use_file")
#define MAIN_USEREG _T("use_reg")
#define MAIN_USEFILE_DEFAULT 0
#define MAIN_MASTERKEY _T("master_key") 
#define MAIN_MASTERKEY_DEFAULT _T("This is a secret key that you should change")
#define MAIN_OBFUSCATED_PASWD _T("obfuscated_password")
#define MAIN_OBFUSCATED_PASWD_DEFAULT _T("")
#define MAIN_SETTINGS_PWD _T("password")
#define MAIN_SETTINGS_PWD_DEFAULT _T("")
#define MAIN_ALLOWED_HOSTS _T("allowed_hosts")
#define MAIN_ALLOWED_HOSTS_DEFAULT _T("127.0.0.1")
#define MAIN_ALLOWED_HOSTS_CACHE _T("cache_allowed_hosts")
#define MAIN_ALLOWED_HOSTS_CACHE_DEFAULT 1
#define MAIN_STRING_LENGTH _T("string_length")
#define MAIN_STRING_LENGTH_DEFAULT 4096
#define MAIN_SHARED_SESSION _T("shared_session")
#define MAIN_SHARED_SESSION_DEFAULT 0


// LOA Config items
#define LUA_SCRIPT_SECTION_TITLE _T("LUA Scripts")


// Main Registry ROOT
#define NS_HKEY_ROOT HKEY_LOCAL_MACHINE
#define NS_REG_ROOT _T("SOFTWARE\\NSClient++")
